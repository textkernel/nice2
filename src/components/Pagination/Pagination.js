import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import styles from './Pagination.scss';
import Button from '../Button';
import PaginationButton from './PaginationButton';

const { block, elem } = bem({
    name: 'Pagination',
    classnames: styles,
    propsToMods: ['align']
});

const defineRange = ({ currentPage, maxPages, totalPages }) => {
    const start = Math.max(2, Math.min(currentPage - 2, totalPages - maxPages + 2));
    const end = Math.min(totalPages, start + maxPages - 2);
    const range = new Array(end - start + 1).fill().map((_, i) => start + i);

    return range;
};

const Pagination = props => {
    const {
        align,
        currentPage,
        maxPages,
        totalPages,
        prevLabel,
        nextLabel,
        lastLabel,
        onClick,
        ...rest
    } = props;

    const currentPageCapped = Math.max(1, Math.min(currentPage, totalPages));
    const isPrevDisabled = currentPageCapped === 1;
    const isNextDisabled = currentPageCapped === totalPages;
    const range = defineRange({
        currentPage: currentPageCapped,
        maxPages,
        totalPages
    });

    const handleClick = e => {
        const {
            dataset: { page = null }
        } = e.target;

        if (!onClick || !page || page > totalPages) {
            return false;
        }

        return onClick(e, page);
    };

    return (
        <nav {...rest} {...block(props)} aria-label="pagination">
            {!!prevLabel && (
                <Button
                    {...elem('button', props)}
                    onClick={handleClick}
                    disabled={isPrevDisabled}
                    context="link"
                    data-page={currentPageCapped - 1}
                    aria-disabled={isPrevDisabled}
                >
                    &lsaquo; {prevLabel}
                </Button>
            )}
            <PaginationButton
                onClick={handleClick}
                isActive={currentPageCapped === 1}
                data-page={1}
            >
                {1}
            </PaginationButton>
            {range[0] > 2 && <div {...elem('gap', props)}>&hellip;</div>}
            {range.map(page => (
                <PaginationButton
                    onClick={handleClick}
                    isActive={page === currentPageCapped}
                    data-page={page}
                    key={`page_${page}`}
                >
                    {page}
                </PaginationButton>
            ))}
            {!!nextLabel && (
                <Button
                    {...elem('button', props)}
                    onClick={handleClick}
                    disabled={isNextDisabled}
                    context="link"
                    data-page={currentPageCapped + 1}
                    aria-disabled={isNextDisabled}
                >
                    {nextLabel} &rsaquo;
                </Button>
            )}
            {!!lastLabel && (
                <Button
                    {...elem('button', props)}
                    onClick={handleClick}
                    disabled={isNextDisabled}
                    context="link"
                    data-page={totalPages}
                    aria-disabled={isNextDisabled}
                >
                    {lastLabel} &raquo;
                </Button>
            )}
        </nav>
    );
};

Pagination.displayName = 'Pagination';

Pagination.propTypes = {
    /** How the buttons should be aligned in the pagination container */
    align: PropTypes.oneOf(['left', 'center', 'right']),
    /** Current page number */
    currentPage: PropTypes.number,
    /** Max. number of pages to list (excluding prev / next) */
    maxPages: PropTypes.number,
    /** Total number of available pages */
    totalPages: PropTypes.number.isRequired,
    /** Label for 'Previous page' button (required for button to show) */
    prevLabel: PropTypes.string,
    /** Label for 'Next page' button (required for button to show) */
    nextLabel: PropTypes.string,
    /** Label for 'Last page' button (required for button to show) */
    lastLabel: PropTypes.string,
    /** Callback function on page / prev/ next click */
    onClick: PropTypes.func
};

Pagination.defaultProps = {
    align: 'center',
    currentPage: 1,
    maxPages: 10,
    prevLabel: null,
    nextLabel: null,
    lastLabel: null,
    onClick: null
};

export default Pagination;
