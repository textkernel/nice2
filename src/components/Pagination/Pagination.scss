.Pagination {
    display: flex;

    &__gap {
        color: var(--color-neutral);
        display: inline-flex;
        align-items: center;
        padding: 0 var(--spacing-3x);
    }
    &--align_left {
        justify-content: flex-start;
    }
    &--align_center {
        justify-content: center;
    }
    &--align_right {
        justify-content: flex-end;
    }
}
