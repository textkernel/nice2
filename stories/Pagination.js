import React from 'react';
import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import { number, select, text, withKnobs } from '@storybook/addon-knobs'; // eslint-disable-line import/no-extraneous-dependencies
import { Pagination } from '@textkernel/oneui';

storiesOf('Pagination', module)
    .addDecorator(withKnobs)
    .add('Pagination', () => (
        <Pagination
            align={select('Button alignment', ['left', 'center', 'right'], 'center')}
            currentPage={number('Current page number', 1)}
            maxPages={number('Max. number of pages to list', 5)}
            totalPages={number('Total number of pages', 25)}
            onClick={(e, page) => {
                console.log(`Requested page ${page}`);
            }}
            prevLabel={text('Previous label', 'Previous')}
            nextLabel={text('Next label', 'Next')}
            lastLabel={text('Last label', 'Last')}
        />
    ));
