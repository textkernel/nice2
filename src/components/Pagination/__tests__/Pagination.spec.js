import React from 'react';
import toJson from 'enzyme-to-json';
import Pagination from '../Pagination';

describe('<Pagination> that renders a pagination component', () => {
    const CURRENT_PAGE = 6;
    const MAX_BUTTONS = 8;
    const onClick = jest.fn();
    let wrapper;

    beforeEach(() => {
        jest.spyOn(console, 'error').mockImplementationOnce(() => {});
        wrapper = mount(
            <Pagination
                currentPage={CURRENT_PAGE}
                maxPageButtons={MAX_BUTTONS}
                totalPages={20}
                onClick={onClick}
                prevLabel="Previous"
                nextLabel="Next"
            />
        );
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('should render correctly with default props', () => {
        wrapper = mount(<Pagination totalPages={20} />);

        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render correctly with all props', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('PaginationButton')).toHaveLength(MAX_BUTTONS);
    });
    describe('click behaviour', () => {
        it('should call onClick with correct paramas when a page button is clicked', () => {
            const e = {
                target: {
                    dataset: {
                        page: 3,
                    },
                },
            };

            wrapper.find('button[data-page=3]').simulate('click', e);

            expect(onClick).toHaveBeenCalledWith(expect.anything(), 3);
        });
        it('should not call onClick when current page button is clicked', () => {
            const e = {
                target: {
                    dataset: {
                        page: CURRENT_PAGE,
                    },
                },
            };

            wrapper.find(`button[data-page=${CURRENT_PAGE}]`).simulate('click', e);

            expect(onClick).not.toHaveBeenCalled();
        });
    });
    describe('data-page property of buttons', () => {
        it('should be set correctly on Prev button', () => {
            expect(
                wrapper
                    .findWhere(el => el.text() === `\u2039 Previous`)
                    .find('button')
                    .prop('data-page')
            ).toBe(CURRENT_PAGE - 1);
        });
        it('should be set correctly on Next button', () => {
            expect(
                wrapper
                    .findWhere(el => el.text() === `Next \u203a`)
                    .find('button')
                    .prop('data-page')
            ).toBe(CURRENT_PAGE + 1);
        });
        it('should be set correctly on "page 1" button', () => {
            expect(
                wrapper
                    .findWhere(el => el.text() === '1')
                    .find('button')
                    .prop('data-page')
            ).toBe(1);
        });
        it('should be set correctly on a page button', () => {
            const pageButton = wrapper.find('PaginationButton').at(2);

            expect(pageButton.text()).toBe('4');
            expect(pageButton.prop('data-page')).toBe(4);
        });
    });
    describe('edge cases', () => {
        it('should throw and error if curernt page is 0', () => {
            expect(() => mount(<Pagination totalPages={20} currentPage={0} />)).toThrow();
        });
        it('should throw and error if curernt page is too large', () => {
            expect(() => mount(<Pagination totalPages={20} currentPage={22} />)).toThrow();
        });
        it('should show only button for page 1 when maxPageButtons is 1 and current page is 1', () => {
            wrapper = mount(<Pagination currentPage={1} maxPageButtons={1} totalPages={20} />);
            const pageButtons = wrapper.find('PaginationButton');

            expect(pageButtons).toHaveLength(1);
            expect(pageButtons.at(0).prop('data-page')).toBe(1);
        });
        it('should show only button for CURRENT_PAGE when maxPageButtons is 1 and current page is not 1', () => {
            wrapper = mount(
                <Pagination currentPage={CURRENT_PAGE} maxPageButtons={1} totalPages={20} />
            );
            const pageButtons = wrapper.find('PaginationButton');

            expect(pageButtons).toHaveLength(1);
            expect(pageButtons.at(0).prop('data-page')).toBe(CURRENT_PAGE);
        });
        it('should show last page when it is the current page as well', () => {
            wrapper = mount(
                <Pagination
                    currentPage={CURRENT_PAGE}
                    maxPageButtons={5}
                    totalPages={CURRENT_PAGE}
                />
            );
            const pageButtons = wrapper.find('PaginationButton');

            expect(pageButtons).toHaveLength(5);
            expect(pageButtons.last().prop('data-page')).toBe(CURRENT_PAGE);
        });
    });
});
