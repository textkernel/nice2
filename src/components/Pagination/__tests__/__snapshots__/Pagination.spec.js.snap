// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination> that renders a pagination component should render a default pagination 1`] = `
<nav
  aria-label="pagination"
  className="Pagination Pagination--align Pagination--align_center"
>
  <Button
    aria-disabled={true}
    className="Pagination__button Pagination__button--align Pagination__button--align_center"
    context="link"
    data-page={0}
    disabled={true}
    isBlock={false}
    isInline={false}
    onClick={[Function]}
    size="normal"
    type="button"
  >
    ‹ 
    Previous
  </Button>
  <PaginationButton
    data-page={1}
    isActive={true}
    onClick={[Function]}
  >
    1
  </PaginationButton>
  <PaginationButton
    data-page={2}
    isActive={false}
    key="page_2"
    onClick={[Function]}
  >
    2
  </PaginationButton>
  <PaginationButton
    data-page={3}
    isActive={false}
    key="page_3"
    onClick={[Function]}
  >
    3
  </PaginationButton>
  <PaginationButton
    data-page={4}
    isActive={false}
    key="page_4"
    onClick={[Function]}
  >
    4
  </PaginationButton>
  <PaginationButton
    data-page={5}
    isActive={false}
    key="page_5"
    onClick={[Function]}
  >
    5
  </PaginationButton>
  <Button
    aria-disabled={false}
    className="Pagination__button Pagination__button--align Pagination__button--align_center"
    context="link"
    data-page={2}
    disabled={false}
    isBlock={false}
    isInline={false}
    onClick={[Function]}
    size="normal"
    type="button"
  >
    Next
     ›
  </Button>
  <Button
    aria-disabled={false}
    className="Pagination__button Pagination__button--align Pagination__button--align_center"
    context="link"
    data-page={20}
    disabled={false}
    isBlock={false}
    isInline={false}
    onClick={[Function]}
    size="normal"
    type="button"
  >
    Last
     »
  </Button>
</nav>
`;
